packages/make-hardener/src/main.js

```diff
@@ -33,113 +59,216 @@ const { ownKeys } = Reflect;
+      answer2 = isFrozen(val) && !onlyNonExtensible.has(val);
+    }
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      if (answer1 !== answer2) {
+        // Not necessarily an error. Might be one of the three safe
+        // exceptions explained in the README.md. We should test these
+        // outliers, because if it is anything else, we may have a
+        // serious bug.
+        console.log(`HARDEN disagreement: ${answer1} vs ${answer2} at ${path}`);
       }
+    }
 
-      function commit() {
-        // todo curried forEach
-        // we capture the real WeakSet.prototype.add above, in case someone
-        // changes it. The two-argument form of forEach passes the second
-        // argument as the 'this' binding, so we add to the correct set.
-        toFreeze.forEach(hardened.add, hardened);
+    return !!(answer1 || answer2);
+  };
+
+  const moreObjectsHardened = toHarden => {
+    if (ENABLE_OLD_HARDENED && ENABLE_ONLY_NON_EXTENSIBLE) {
+      // TODO What consistency check should we put here?
+    }
+    if (ENABLE_OLD_HARDENED) {
+      toHarden.forEach(weakSetAdd, oldHardened);
+    }
+    if (ENABLE_ONLY_NON_EXTENSIBLE) {
+      toHarden.forEach(weakSetDelete, onlyNonExtensible);
+    }
+  };
+
+  /**
+   * @template T
+   * @param {T} root
+   * @returns {T}
+   */
+  const harden = root => {
+    const toHarden = new Set();
+    const paths = new WeakMap();
+
+    // If val is something we should be freezing but aren't yet,
+    // add it to toFreeze.
+    /**
+     * @param {any} val
+     * @param {string} [path]
+     */
+    function enqueue(val, path = undefined) {
+      if (Object(val) !== val) {
+        // ignore primitives
+        return;
+      }
+      const type = typeof val;
+      if (type !== 'object' && type !== 'function') {
+        // future proof: break until someone figures out what it should do
+        throw new TypeError(`Unexpected typeof: ${type}`);
+      }
+      if (isHardened(val, path) || toHarden.has(val)) {
+        // Ignore if this is an exit, or we've already visited it
+        return;
       }
+      // console.log(`adding ${val} to toFreeze`, val);
+      toHarden.add(val);
+      paths.set(val, path);
+    }
+
+    /**
+     * @param {any} obj
+     */
+    function freezeAndTraverse(obj) {
+      // Now freeze the object to ensure reactive
+      // objects such as proxies won't add properties
+      // during traversal, before they get frozen.
 
-      enqueue(root);
-      dequeue();
-      // console.log("toFreeze set:", toFreeze);
-      commit();
+      // Object are verified before being enqueued,
+      // therefore this is a valid candidate.
+      // Throws if this fails (strict mode).
+      patchedObjectFreeze(obj);
 
-      return root;
-    },
+      // we rely upon certain commitments of Object.freeze and proxies here
+
+      // get stable/immutable outbound links before a Proxy has a chance to do
+      // something sneaky.
```

Comment:
I note that the proxy rules also say things about "after you've gotten an answer to isExtensible then it can't change".  Should this also retrieve isExtensible?